## Business Analytics - Programming Task 1 {-}


---

**Assignment Points**: 100  
**Due Date**: Friday Week 6 (4 April 2025) @ 11.59pm  
**Submission**: Provide your answers in this Jupypter notebook and submit it via iLearn link  



## **Predictive Analysis of Credit Card Defaults**

### Objective: 
- This assignment focuses on a dataset of customers' default payments.
- The primary goal is to predict which credit card clients are likely to default using various data mining methods.

### Background: 
Traditional risk management models classify clients as either credible or not credible based on their likelihood of default. This project aims to refine this classification by identifying specific individuals who are likely to default, enhancing the precision of credit risk assessments.

Target variable
- default.payment.next.month: Default payment (1=yes, 0=no)

The dataset contains the following features

1. ID: ID of each client
2. LIMIT_BAL: Amount of given credit in dollars (includes individual and family/supplementary credit
3. SEX: Gender (1=male, 2=female)
4. EDUCATION: (1=graduate school, 2=university, 3=high school, 4=others, 5=unknown, 6=unknown)
5. MARRIAGE: Marital status (1=married, 2=single, 3=others)
6. AGE: Age in years
7. PAY_0: Repayment status in September (-1=pay duly, 1=payment delay for one month, 2=payment delay for two months, ... 8=payment delay for eight months, 9=payment delay for nine months and above)
8. PAY_2: Repayment status in August (scale same as above)
9. PAY_3: Repayment status in July, (scale same as above)
10. PAY_4: Repayment status in June (scale same as above)
11. PAY_5: Repayment status in May (scale same as above)
12. PAY_6: Repayment status in April (scale same as above)
13. BILL_AMT1: Amount of bill statement in September (dollars)  
14. BILL_AMT2: Amount of bill statement in August (dollars)  
15. BILL_AMT3: Amount of bill statement in July (dollars)  
16. BILL_AMT4: Amount of bill statement in June (dollars)  
17. BILL_AMT5: Amount of bill statement in May (dollars)  
18. BILL_AMT6: Amount of bill statement in April (dollars)   
19. PAY_AMT1: Amount of previous payment in September (dollars)  
20. PAY_AMT2: Amount of previous payment in August (dollars)  
21. PAY_AMT3: Amount of previous payment in July (dollars)   
22. PAY_AMT4: Amount of previous payment in June (dollars)  
23. PAY_AMT5: Amount of previous payment in May (dollars)   
24. PAY_AMT6: Amount of previous payment in April (dollars)  



---

**Problem 1** - Reading the dataset (Total Marks: 20)



**Q1**. Create a pandas dataframe contining the first 10,000 rows from the credit card dataset provided in the **assignment_data** folder 
- Delete 'ID' column
- Print `info()` of the dataframe 

(5 marks) 




In [5]:
# ---- provide your code here -----
import pandas as pd

# Load the dataset with the first 10,000 rows.
df = pd.read_excel('assignment_data/credit_data.xlsx', nrows=10000)

# Drop the 'ID' column.
df.drop(columns=['ID'], inplace=True)

# Print the info of the DataFrame to inspect its structure.
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   LIMIT_BAL                   10000 non-null  int64  
 1   SEX                         9617 non-null   float64
 2   EDUCATION                   9617 non-null   float64
 3   MARRIAGE                    9617 non-null   float64
 4   AGE                         9617 non-null   float64
 5   PAY_0                       9617 non-null   float64
 6   PAY_2                       9617 non-null   float64
 7   PAY_3                       9617 non-null   float64
 8   PAY_4                       9617 non-null   float64
 9   PAY_5                       9617 non-null   float64
 10  PAY_6                       9642 non-null   float64
 11  BILL_AMT1                   9642 non-null   float64
 12  BILL_AMT2                   9642 non-null   float64
 13  BILL_AMT3                   9642

<hr style="width:25%;margin-left:0;"> 

**Q2**. List which **features** are *numeric*, *ordinal*, and *nominal* variables, and how many features of each kind there are in the dataset.
To answer this question 

- Find the definitions of numeric, ordinal and nominal variables in the course material    
- Carefully consider what values each feature can take as well as the output of `df.info()`. 

Your answer should be written up in Markdown and include:
1) A table listing all the features present in the dataset and their type (fill out the table template provided below) and
2) A brief description of the contents of the table.

|Variable Kind|Number of Features|Feature Names
| --- | --- | --- |
| Numeric | some number | some text |
| some text  | some number | some text |
| some text  | some number | some text |


(10 marks)

| Variable Kind | Number of Features | Feature Names |
|---------------|--------------------|----------------|
| Numeric       | 14                 | LIMIT_BAL, AGE, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6 |
| Ordinal       | 7                  | EDUCATION, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6 |
| Nominal       | 3                  | SEX, MARRIAGE, default payment next month |

---- provide your text answer here ----

**Description:**

**Numeric Variables:**
These features represent measurable quantities (e.g., credit limit, age, bill amounts, payment amounts) that can be used in mathematical operations such as addition, subtraction, and averaging.

**Ordinal Variables:**
These variables (e.g., education levels and repayment statuses) have a clear ranking or sequence, even though the exact distance between categories may not be consistent. For instance, different levels of education imply an ordered progression, and higher PAY_x values indicate longer payment delays.

**Nominal Variables:**
These are categorical variables without an inherent order. In this dataset, SEX, MARRIAGE, and default.payment.next.month fall into this category, as they simply distinguish between groups (e.g., male/female, married/single, default/non-default) without any ranked relationship.

<hr style="width:25%;margin-left:0;"> 

**Q3.** Missing Values. 

- Print out the number of missing values for each variable in the dataset and comment on your findings.

(5 marks)

In [10]:
# ---- provide your code here -----
# Check for missing values in each column of the dataframe
missing_values = df.isnull().sum()
print(missing_values)

LIMIT_BAL                       0
SEX                           383
EDUCATION                     383
MARRIAGE                      383
AGE                           383
PAY_0                         383
PAY_2                         383
PAY_3                         383
PAY_4                         383
PAY_5                         383
PAY_6                         358
BILL_AMT1                     358
BILL_AMT2                     358
BILL_AMT3                     358
BILL_AMT4                     358
BILL_AMT5                     368
BILL_AMT6                     368
PAY_AMT1                      368
PAY_AMT2                      368
PAY_AMT3                       10
PAY_AMT4                       10
PAY_AMT5                      290
PAY_AMT6                      290
default payment next month      0
dtype: int64


---- provide your text answer here ----

**Insights:**

- **Complete Variables:** LIMIT_BAL and the target variable (default payment next month) have no missing values, indicating complete data availability.

- **Moderate Missingness:** Several features such as SEX, EDUCATION, MARRIAGE, AGE, and most of the repayment status (PAY_0 to PAY_6) and billing amount (BILL_AMT1 to BILL_AMT6) variables have a noticeable but still manageable amount of missing values.

- **Low Missingness:** PAY_AMT3 and PAY_AMT4 have very few missing entries compared to other variables, indicating higher completeness.

Overall, while the dataset is mostly complete, addressing the existing missing values through appropriate methods (e.g., imputation or removal) will be essential to maintain data quality and ensure reliable analysis results.

---
---

**Problem 2.** Cleaning data and dealing with categorical features (Total Marks: 40)


**Q1.** 

- Use an appropriate `pandas` function to impute missing values using one of the following two strategies: `mean` and `mode`. (10 marks)
- Take into consideration the type of each variable (as in Q2 above) and the best practices we discussed in class/lecture notes
- Explain what data imputation is, how you have done it here, and what decisions you had to make. (5 marks)


(Total: 15 marks)

In [14]:
# ---- provide your code here -----
import pandas as pd

# Define features based on the classification:
numeric_features = [
    'LIMIT_BAL', 'AGE',
    'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6',
    'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6'
]

ordinal_features = ['EDUCATION', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6']
nominal_features = ['SEX', 'MARRIAGE', 'default payment next month']

# Impute missing values in numeric columns using the mean
df[numeric_features] = df[numeric_features].fillna(df[numeric_features].mean())

# Impute missing values in ordinal and nominal columns using the mode
for col in ordinal_features + nominal_features:
    df[col] = df[col].fillna(df[col].mode()[0])

---- provide your text answer here ----

**Data Imputation Explanation:**

Data imputation is the process of replacing missing values in a dataset with estimated values so that the dataset becomes complete and can be used for further analysis or modeling. This is important because missing data can lead to biased results or errors when training predictive models.

**How It Was Done Here:**

- **Numeric Variables:**
For continuous variables (LIMIT_BAL, AGE, bill amounts, and payment amounts), the missing values were replaced using the mean of the available data. This is appropriate because the mean provides a representative value for variables that can be added, subtracted, or averaged.

- **Ordinal and Nominal Variables:**
For categorical variables, both ordinal (EDUCATION and repayment status features such as PAY_0 to PAY_6) and nominal (SEX, MARRIAGE, and default payment next month), the missing values were replaced using the mode (the most frequent value). This decision is based on the fact that for categorical data, computing an average is not meaningful, and using the mode maintains the most common category.

**Decisions Made:**

- **Method Choice:**
The decision to use mean for numeric variables and mode for categorical ones aligns with best practices taught in class. This ensures that the imputation respects the nature of the data.

- **Feature Segregation:**
The dataset was first divided into numeric, ordinal, and nominal features. This classification guided which imputation strategy to use for each group, ensuring that the replacement value is appropriate for the type of data.

- **Data Integrity:**
By selecting these methods, we aim to minimise the distortion of the original data distribution, which is crucial for maintaining the quality and reliability of any subsequent analysis or model training.

<hr style="width:25%;margin-left:0;"> 

**Q2**. 
- Print `value_counts()` of the 'SEX' column and add a dummy variable named 'SEX_FEMALE' to `df` using `get_dummies()` (3 marks)
- Carefully explain what the values of the new variable 'SEX_FEMALE' mean (2 mark)
- Make sure the variable 'SEX' is deleted from the original dataframe   

(Total: 5 marks)  

In [17]:
# ---- provide your code here -----
# Print the counts of each value in the 'SEX' column
print(df['SEX'].value_counts())

# Create dummy variables for the 'SEX' column using get_dummies()
sex_dummies = pd.get_dummies(df['SEX'], prefix='SEX')

# In the dataset, 'SEX' is coded as:
#   1 = Male, 2 = Female
# The dummy variable for females is originally named 'SEX_2.0'. So I renamed it to 'SEX_FEMALE' as instructed
sex_dummies.rename(columns={'SEX_2.0': 'SEX_FEMALE'}, inplace=True)

# Add the 'SEX_FEMALE' column to the dataframe
df = pd.concat([df, sex_dummies['SEX_FEMALE']], axis=1)

# Remove the original 'SEX' column
df.drop(columns=['SEX'], inplace=True)

# To check 'SEX' column has been deleted
df.info()

SEX
2.0    6162
1.0    3838
Name: count, dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   LIMIT_BAL                   10000 non-null  int64  
 1   EDUCATION                   10000 non-null  float64
 2   MARRIAGE                    10000 non-null  float64
 3   AGE                         10000 non-null  float64
 4   PAY_0                       10000 non-null  float64
 5   PAY_2                       10000 non-null  float64
 6   PAY_3                       10000 non-null  float64
 7   PAY_4                       10000 non-null  float64
 8   PAY_5                       10000 non-null  float64
 9   PAY_6                       10000 non-null  float64
 10  BILL_AMT1                   10000 non-null  float64
 11  BILL_AMT2                   10000 non-null  float64
 12  BILL_AMT3                   10000 n

---- provide your text answer here ----

**'SEX_FEMALE' Variable Explanation:**

The new variable 'SEX_FEMALE' is a binary indicator derived from the original 'SEX' column using dummy encoding. Here are what the values represent:

- 1: Indicates that the individual is female.

- 0: Indicates that the individual is not female (i.e., the individual is male).

This transformation simplifies the gender information into a clear and easy-to-use format for further analysis or modeling.

<hr style="width:25%;margin-left:0;"> 

**Q3**. Print `value_counts()` of the 'MARRIAGE' column and *carefully* comment on what you notice in relation to the definition of this variable. 

(Total: 5 marks) 

In [20]:
# ---- provide your code here -----
# Print value counts for the 'MARRIAGE' column
print(df['MARRIAGE'].value_counts())

MARRIAGE
2.0    5518
1.0    4380
3.0      82
0.0      20
Name: count, dtype: int64


---- provide your text answer here ----

**Expected Categories:**

According to the variable definition, **MARRIAGE** should have:

- 1: Married

- 2: Single

- 3: Others


**Observations:**

- The majority of records are coded as **2.0 (Single)** and **1.0 (Married)**, which is in line with the expected values.

- A small number of records (82) are classified as **3.0 (Others).**

- **Unexpected Category:** There are **20 records** with a value of **0.0**, which does not match the defined categories.


**Implications:**

The presence of a **0.0** value suggests that there might be data entry errors, misclassifications, or perhaps an undocumented category (such as unknown or missing) in the dataset. This anomaly should be investigated further to determine if these values need correction or exclusion during preprocessing.

<hr style="width:25%;margin-left:0;"> 

**Q4**. 

- Apply `get_dummies()` to 'MARRIAGE' feature and add dummy variables 'MARRIAGE_MARRIED', 'MARRIAGE_SINGLE', 'MARRIAGE_OTHER' to `df`. (5 marks)   
- *Carefully consider* how to allocate all the values of 'MARRIAGE' across these 3 newly created features (5 marks)
    - Do not delete observations 
    - Do not assume that the anomaly are missing observations      
    - Explain what decision you made
- Make sure that 'MARRIAGE' is deleted from `df`   

(Total: 10 marks)   

In [23]:
# ---- provide your code here -----
# Create dummy variables for the 'MARRIAGE' feature using get_dummies()
marriage_dummies = pd.get_dummies(df['MARRIAGE'], prefix='MARRIAGE')

# Map the dummy columns to the target variable names:
# According to the definition:
#   1.0 -> Married  -> MARRIAGE_MARRIED
#   2.0 -> Single   -> MARRIAGE_SINGLE
#   3.0 -> Others   -> MARRIAGE_OTHER

# However, the dataset also contains a value 0.0. 
# So I decided to allocate the 0.0 values to the "Others" category,
# since they do not clearly fit into "Married" (1.0) or "Single" (2.0).

df['MARRIAGE_MARRIED'] = marriage_dummies.get('MARRIAGE_1.0', 0)
df['MARRIAGE_SINGLE']  = marriage_dummies.get('MARRIAGE_2.0', 0)
df['MARRIAGE_OTHER']   = marriage_dummies.get('MARRIAGE_3.0', 0) + marriage_dummies.get('MARRIAGE_0.0', 0)

# Remove the original 'MARRIAGE' column from the dataframe
df.drop(columns=['MARRIAGE'], inplace=True)

# To check 'MARRIAGE' column has been deleted
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 26 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   LIMIT_BAL                   10000 non-null  int64  
 1   EDUCATION                   10000 non-null  float64
 2   AGE                         10000 non-null  float64
 3   PAY_0                       10000 non-null  float64
 4   PAY_2                       10000 non-null  float64
 5   PAY_3                       10000 non-null  float64
 6   PAY_4                       10000 non-null  float64
 7   PAY_5                       10000 non-null  float64
 8   PAY_6                       10000 non-null  float64
 9   BILL_AMT1                   10000 non-null  float64
 10  BILL_AMT2                   10000 non-null  float64
 11  BILL_AMT3                   10000 non-null  float64
 12  BILL_AMT4                   10000 non-null  float64
 13  BILL_AMT5                   1000

---- provide your text answer here ----

**Explanation:**

**1. Creating Dummy Variables:**
I used pd.get_dummies() to transform the 'MARRIAGE' column into dummy variables. This creates separate columns for each unique value (MARRIAGE_0.0, MARRIAGE_1.0, MARRIAGE_2.0, and MARRIAGE_3.0).

**2. Mapping to Desired Dummy Names:**

- MARRIAGE_MARRIED: I mapped observations with a value of 1.0 to this dummy variable.

- MARRIAGE_SINGLE: I mapped observations with a value of 2.0 to this dummy variable.

- MARRIAGE_OTHER: I mapped observations with a value of 3.0 to this variable.


**Handling the Anomaly (0.0):**
The dataset also contains 0.0, which is not part of the original definition. Instead of discarding or treating these as missing, I chose to add them into "Others" category. This decision is based on the reasoning that 0.0 does not clearly indicate either Married or Single, so it is most appropriately grouped under "Others."


**3. Finalising the DataFrame:**
After creating and mapping the dummy variables, I removed the original 'MARRIAGE' column from the dataframe. This ensures our dataframe now only contains the new three clearly defined dummy variables while retaining all observations.

<hr style="width:25%;margin-left:0;"> 

**Q5**. In the column 'EDUCATION', convert the values {0, 5, 6} to the value 4. 

(Total: 5 marks)  

In [26]:
# ---- provide your code here -----
# Replace values 0, 5, and 6 in the 'EDUCATION' column with 4
df['EDUCATION'] = df['EDUCATION'].replace({0: 4, 5: 4, 6: 4})

---
---

**Problem 3** Preparing X and y arrays (Total Marks: 10)

**Q1**. 

- Create a numpy array `y` from the first 7,000 observations of `default payment next month` column from `df` (2.5 marks)   
- Create a numpy array `X`  from the first 7,000 observations of all the remaining variables in `df` (2.5 marks)   

(Total: 5 Marks)

In [29]:
# ---- provide your code here -----
import numpy as np

# Create the target array y from the first 7000 observations of the target variable
y = df['default payment next month'].values[:7000]

# Create the feature array X from the first 7000 observations of all remaining variables
X = df.drop(columns=['default payment next month']).values[:7000]

<hr style="width:25%;margin-left:0;"> 

**Q2**. 

- Use an appropriate `sklearn` library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% train and 30% test datasets (2.5 marks) 
    - Set random_state to 31 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
- Standardise the data to mean zero and variance one using an approapriate `sklearn` library (2.5 marks)   


(Total: 5 marks) 

In [31]:
# ---- provide your code here -----
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Split the data into training and test sets (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=31, stratify=y
)

# Standardise the features: mean=0 and variance=1
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)  # Fit and transform on training data
X_test_scaled = scaler.transform(X_test)        # Transform test data using the same scaler

---
---

**Problem 4**. Training Models and Interpretation (Total Marks: 30)


**Q1**. 

- Train one linear classifier we studied in class using standardised data (6 marks)
- Compute and print training and test dataset accuracies (4 marks)

(Total: 10 marks)   

In [34]:
# ---- provide your code here -----
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initialise the Logistic Regression model
log_reg = LogisticRegression(random_state=31, max_iter=1000)

# Train the model using the standardised training data
log_reg.fit(X_train_scaled, y_train)

# Predict on the training and test sets
y_train_pred = log_reg.predict(X_train_scaled)
y_test_pred = log_reg.predict(X_test_scaled)

# Compute accuracies for training and test sets
train_accuracy = accuracy_score(y_train, y_train_pred)
test_accuracy = accuracy_score(y_test, y_test_pred)

# Print the results
print("Training Accuracy: {:.2f}%".format(train_accuracy * 100))
print("Test Accuracy: {:.2f}%".format(test_accuracy * 100))

Training Accuracy: 82.06%
Test Accuracy: 80.52%


<hr style="width:25%;margin-left:0;"> 

**Q2.**

- Train one nonlinear classifier we studied in class on the same dataset (6 marks)
- Compute and print training and test dataset accuracies (4 marks)


(Total: 10 marks)  

In [36]:
# ---- provide answer here -----
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Initialise the RandomForestClassifier with a fixed random state for reproducibility
rf_clf = RandomForestClassifier(random_state=31)

# Train the model using the standardised training data
rf_clf.fit(X_train_scaled, y_train)

# Make predictions on both the training and test datasets
y_train_pred_rf = rf_clf.predict(X_train_scaled)
y_test_pred_rf = rf_clf.predict(X_test_scaled)

# Calculate the accuracy for the training and test datasets
train_accuracy_rf = accuracy_score(y_train, y_train_pred_rf)
test_accuracy_rf = accuracy_score(y_test, y_test_pred_rf)

# Print the results
print("Random Forest Training Accuracy: {:.2f}%".format(train_accuracy_rf * 100))
print("Random Forest Test Accuracy: {:.2f}%".format(test_accuracy_rf * 100))

Random Forest Training Accuracy: 99.98%
Random Forest Test Accuracy: 81.05%


<hr style="width:25%;margin-left:0;"> 

**Q3**. 

- Comment on the accuracy results obtained from the two classifiers (6 marks)
- Based on our investigation into credit card default predictions in this assignment, which model would you recommend? (4 marks)


(Total: 10 marks)     


---- provide your text answer here ----

**Comments on Accuracy Results**

**1. Logistic Regression (Linear Classifier):**

- Training Accuracy: 82.06%

- Test Accuracy: 80.52%

**Interpretation:**

The logistic regression model shows consistent performance between the training and test sets. The small difference in accuracy (about 1.5 percentage points) indicates that the model generalises well and is not overfitting.

**2. Random Forest (Non-Linear Classifier):**

- Training Accuracy: 99.98%

- Test Accuracy: 81.05%

**Interpretation:**

Although the Random Forest model achieves nearly perfect accuracy on the training data, its test accuracy is very similar to the logistic regression results. This large gap between training and test accuracy suggests that the Random Forest model is overfitting the training data by memorising the training examples rather than learning generalisable patterns.

**3. Model Recommendation**

Based on my investigation into credit card default predictions, I would recommend using the Logistic Regression model.

**Reasons:**

- **Generalisation:** Logistic Regression provides a more reliable and robust performance on unseen data, with similar test accuracy to the Random Forest but without the risk of overfitting.

- **Interpretability:** The linear nature of logistic regression makes it easier to interpret and explain, which is valuable in a business analytics context.

- **Simplicity:** A simpler model that performs comparably on the test set is preferable, especially when it avoids overfitting.

---
---

## Marking Criteria

To achieve a perfect score, your solutions must adhere to the criteria outlined below:

- Ensure that all numerical answers are accurate.
- Utilize the exact Python functions and libraries specified within the assignment instructions.
- For any written responses, provide accurate information, articulated in clear, complete sentences.
- Do not add extra cells beyond what is provided in the notebook.
- Do not print output with your code unless explicitly instructed to do so.
- Maintain a clean and organised notebook layout that is easy to follow.
- Marks will be deducted for not following the above instructions.
    
---
---